fido/Crypto/Fido2/Model/JavaScript.hs

```diff
@@ -0,0 +1,251 @@
+-- field, but only values @"foo"@ and @"bar"@ are possible, the type is still encoded as a generic 'DOMString'.
+-- This allows us to match the specification very closely, deferring decoding of these values to another module.
+-- This module also implements 'Aeson.FromJSON' and 'Aeson.ToJSON' instances of its types, which are compatible with
+-- [webauthn-json](https://github.com/github/webauthn-json)'s JSON schema.
+--
+-- The defined types are
+--
+-- - 'PublicKeyCredentialCreationOptions' and all its subtypes. Passed as the
+--   [publicKey](https://www.w3.org/TR/webauthn-2/#dom-credentialcreationoptions-publickey) field to the
+--   [create()](https://w3c.github.io/webappsec-credential-management/#dom-credentialscontainer-create) method
+--   in step 2 of [§ 7.1 Registering a New Credential](https://www.w3.org/TR/webauthn-2/#sctn-registering-a-new-credential)
+-- - 'PublicKeyCredentialRequestOptions' and all its subtypes. Passed as the
+--   [publicKey](https://www.w3.org/TR/webauthn-2/#dom-credentialrequestoptions-publickey) field to the
+--   [get()](https://w3c.github.io/webappsec-credential-management/#dom-credentialscontainer-get) method
+--   in step 2 of [§ 7.2 Verifying an Authentication Assertion](https://www.w3.org/TR/webauthn-2/#sctn-verifying-assertion)
+-- - @'PublicKeyCredential' response@ and all its subtypes. Responses of the
+--   [create()](https://w3c.github.io/webappsec-credential-management/#dom-credentialscontainer-create) (in which case @response ~ 'AuthenticatorAttestationResponse'@) and
+--   [get()](https://w3c.github.io/webappsec-credential-management/#dom-credentialscontainer-get) (in which case @response ~ 'AuthenticatorAssertionResponse'@ methods.
+module Crypto.Fido2.Model.JavaScript
+  ( -- * Top-level types
+    PublicKeyCredentialCreationOptions (..),
+    PublicKeyCredentialRequestOptions (..),
+    CreatedPublicKeyCredential,
+    RequestedPublicKeyCredential,
+
+    -- * Nested types
+    PublicKeyCredential (..),
+    AuthenticatorAttestationResponse (..),
+    AuthenticatorAssertionResponse (..),
+    PublicKeyCredentialRpEntity (..),
+    PublicKeyCredentialUserEntity (..),
+    PublicKeyCredentialParameters (..),
+    COSEAlgorithmIdentifier,
+    PublicKeyCredentialDescriptor (..),
+    AuthenticatorSelectionCriteria (..),
+
+    -- * JavaScript-builtin types
+    DOMString,
+    UnsignedLong,
+    BufferSource (..),
+    ArrayBuffer,
+  )
+where
+
+import qualified Data.Aeson as Aeson
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Base64.URL as Base64
+import Data.Int (Int32)
+import Data.Map (Map)
+import Data.Text (Text)
+import qualified Data.Text.Encoding as Text
+import Data.Word (Word32)
+import Deriving.Aeson
+  ( CustomJSON (CustomJSON),
+    FieldLabelModifier,
+    OmitNothingFields,
+    Rename,
+  )
+import GHC.Generics (Generic)
+
+-- TODO: It's probably not okay for both DOMString and USVString to be the
+-- same type, figure out how they need to differ
```

Comment:
I believe the only restriction is that a USVString is guaranteed not have any surrogate code points. Because it does not contain any surrogate code points a [Word16] is a valid representation of USVString, which is not true for DOMString, because it can contain surrogate code points.

Also, we should probably make it clear that this modules requires the text package version to be < 2.0